internal/daemon/controller/handlers/workers/worker_service_test.go

```diff
@@ -1791,3 +1850,157 @@ func TestService_RemoveWorkerTags(t *testing.T) {
+			},
+			wantErr:         true,
+			wantErrIs:       handlers.ApiErrorWithCode(codes.InvalidArgument),
+			wantErrContains: "Must be 'global'",
+		},
+		{
+			name:    "success",
+			scopeId: scope.Global.String(),
+			req: &pbs.ReadCertificateAuthorityRequest{
+				ScopeId: "global",
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got, err := testSrv.ReadCertificateAuthority(auth.DisabledAuthTestContext(iamRepoFn, tc.scopeId), tc.req)
+			if tc.wantErr {
+				require.Error(err)
+				assert.Nil(got)
+				if tc.wantErrIs != nil {
+					assert.ErrorIs(err, tc.wantErrIs)
+				}
+				if tc.wantErrContains != "" {
+					assert.Contains(err.Error(), tc.wantErrContains)
+				}
+				return
+			}
+			require.NoError(err)
+			require.NotNil(got)
+		})
+	}
+}
+
+func TestReinitializeCertificateAuthority(t *testing.T) {
+	t.Parallel()
+	require, assert := require.New(t), assert.New(t)
+	ctx := context.Background()
+	wrapper := db.TestWrapper(t)
+	conn, _ := db.TestSetup(t, "postgres")
+	rw := db.New(conn)
+	kmsCache := kms.TestKms(t, conn, wrapper)
+
+	iamRepo := iam.TestRepo(t, conn, wrapper)
+	iamRepoFn := func() (*iam.Repository, error) {
+		return iamRepo, nil
+	}
+
+	repoFn := func() (*server.Repository, error) {
+		return server.NewRepository(rw, rw, kmsCache)
+	}
+
+	workerAuthRepo, err := server.NewRepositoryStorage(ctx, rw, rw, kmsCache)
+	require.NoError(err)
+	workerAuthRepoFn := func() (*server.WorkerAuthRepositoryStorage, error) {
+		return workerAuthRepo, nil
+	}
+
+	// Store CA and check that initial version updates
+	_, err = rotation.RotateRootCertificates(ctx, workerAuthRepo)
+	require.NoError(err)
+
+	testSrv, err := NewService(ctx, repoFn, iamRepoFn, workerAuthRepoFn)
+	require.NoError(err, "Error when getting new worker service.")
+
+	tests := []struct {
+		name            string
+		scopeId         string
+		req             *pbs.ReinitializeCertificateAuthorityRequest
+		wantErr         bool
+		wantErrIs       error
+		wantErrContains string
+	}{
+		{
+			name:    "invalid-id",
+			scopeId: scope.Global.String(),
+			req: &pbs.ReinitializeCertificateAuthorityRequest{
+				ScopeId: "bogus",
+			},
+			wantErr:         true,
+			wantErrIs:       handlers.ApiErrorWithCode(codes.InvalidArgument),
+			wantErrContains: "Must be 'global'",
+		},
+		{
+			name:    "success",
+			scopeId: scope.Global.String(),
+			req: &pbs.ReinitializeCertificateAuthorityRequest{
+				ScopeId: "global",
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got, err := testSrv.ReinitializeCertificateAuthority(auth.DisabledAuthTestContext(iamRepoFn, tc.scopeId), tc.req)
+			if tc.wantErr {
+				require.Error(err)
+				assert.Nil(got)
+				if tc.wantErrIs != nil {
+					assert.ErrorIs(err, tc.wantErrIs)
+				}
+				if tc.wantErrContains != "" {
+					assert.Contains(err.Error(), tc.wantErrContains)
+				}
+				return
+			}
+			require.NoError(err)
+			require.NotNil(got)
```

Comment:
It'd probably be good to read and store the public keys from before the reinitialize call (just by running a Read on the CA), then compare to ensure that both public keys are now different.

Reply:
Oh, I see that in the test below. Never mind!